### ISSUE TITLE
Where should the folder for all-purpose CNS scripts be placed?

### ISSUE THREAD

**Issue by joaomcteixeira** [haddocking/haddock3#45]
If we are to store CNS scripts that (can) serve several HADDOCK3 modules in a separate folder, where should that folder be?

possibilities:

* `src/haddock/cns`, this would imply removing the `py` files from here and placing them in `libs.libcns.py`.
* `src/cns`, `src/haddock/__init__.py` would have a pointer to this folder.
* *others?*

**Comment by rvhonorato** [comment]
We could just have a snapshot of the current CNS protocols in the production, and have them here for safekeeping, but over time it will keep diverging from production and we'd need to refer to it and keep updating it, so we might as well just use the production repository.

For now why don't we keep the scripts can be added in the current (temporary) module-based `cns/` folder?

**Comment by amjjbonvin** [comment]
Unavoidable the cns scripts will diverge

Also if we want to modularise things this should also apply to the cns code

**Comment by joaomcteixeira** [comment]
Following today's discussion, we decided that haddock3 modules/actions will be simplified from their state in haddock2. That is, modules in haddock2 for which internal workflow is controlled by independent non-overlapping parameters will be split into separate modules containing only combinable subroutines or routines part of a strategy selection point. So, for example, if module M can use subroutine A or B but cannot use nor A nor B if C is chosen (and vice-versa), there should be M1 with A and B and M2 with C - two different modules M1 and M2.

M1 and M2 will have their size and dependencies considerably reduced in the above's paradigm, but they will also duplicate common code.

Following our [comments on #40](https://github.com/haddocking/haddock3/issues/40#issuecomment-888663432), `modules/CATEGORY/module1/cns` should contain the CNS code for M1. If `module1` and `module2` share CNS code, that code should go to a common CNS folder so that both modules can pull the code from there.

**What if an advanced user wants to change the shared code only for one of the modules?**

HADDOCK3 python shell should search for CNS code first locally inside the module's folder. If not found, search globally in the general CNS folder. Also, we might think about the option of giving CNS code blocks as run parameters. This could facilitate modifying a module without going to the HADDOCK3 source directory. Users could interchange configs with CNS code blocks without interfering with the haddock3 source code. This is an advanced feature but not hard to implement. To implement it, we need to pull mandatory CNS script names from the modules signatures and convert them into parameters as discussed [for the config generator](https://github.com/haddocking/haddock3/issues/49).

**Comment by rvhonorato** [comment]
> What if an advanced user wants to change the shared code only for one of the modules?

Each of the modules provided by us should be consolidated, meaning any modification to it outside the specification of parameters falls into user-development. In a later stage there will be guides on how to create and modify the modules from a developer perspective. 

I don't think we should go all the way to implement a CNS developer interface, this is a bit outside the scope.

**Comment by joaomcteixeira** [comment]
> I don't think we should go all the way to implement a CNS developer interface, this is a bit outside the scope.

I agree. I did not mean to implement a CNS developer interface. I was thinking on what we discussed about prioritizing the source of the CNS scripts (local vs. global). At the end I though about passing CNS scripts as run parameters thinking about a very advanced user and avoiding the need to place modified CNS modules inside the source folders. But it was just a wondering idea.
